exporter/sumologicexporter/sender.go

```diff
@@ -0,0 +1,258 @@
+func (s *sender) logToText(record pdata.LogRecord) (string, error) {
+	return record.Body().StringVal(), nil
+}
+
+// logToText converts LogRecord to a json line, returns it and error eventually
+func (s *sender) logToJSON(record pdata.LogRecord) (string, error) {
+	data := s.filter.filterOut(record.Attributes())
+	data[logKey] = record.Body().StringVal()
+
+	nextLine, err := json.Marshal(data)
+	if err != nil {
+		return "", err
+	}
+
+	return bytes.NewBuffer(nextLine).String(), nil
+}
+
+// sendLogs sends log records from the buffer formatted according
+// to configured LogFormat and as the result of execution
+// returns array of records which has not been sent correctly and error
+func (s *sender) sendLogs(flds fields) ([]pdata.LogRecord, error) {
+	var (
+		body           strings.Builder
+		errs           []error
+		droppedRecords []pdata.LogRecord
+		currentRecords []pdata.LogRecord
+		formattedLine  string
+		err            error
+	)
+
+	for _, record := range s.buffer {
+		switch s.config.LogFormat {
+		case TextFormat:
+			formattedLine, err = s.logToText(record)
+		case JSONFormat:
+			formattedLine, err = s.logToJSON(record)
+		default:
+			droppedRecords = append(droppedRecords, record)
+			errs = append(errs, errors.New("unexpected log format"))
+			continue
+		}
+
+		if err != nil {
+			droppedRecords = append(droppedRecords, record)
+			errs = append(errs, err)
+			continue
+		}
+
+		ar := s.appendAndSend(formattedLine, LogsPipeline, &body, flds)
+		if ar.err != nil {
+			errs = append(errs, ar.err)
+			if ar.sent {
+				droppedRecords = append(droppedRecords, currentRecords...)
+			}
+
+			if !ar.appended {
+				droppedRecords = append(droppedRecords, record)
+			}
+		}
+
+		// If data was sent, cleanup the currentTimeSeries counter
+		if ar.sent {
+			currentRecords = currentRecords[:0]
+		}
+
+		// If log has been appended to body, increment the currentTimeSeries
+		if ar.appended {
+			currentRecords = append(currentRecords, record)
+		}
+	}
+
+	if err := s.send(LogsPipeline, strings.NewReader(body.String()), flds); err != nil {
+		errs = append(errs, err)
+		droppedRecords = append(droppedRecords, currentRecords...)
+	}
+
+	if len(errs) > 0 {
+		return droppedRecords, componenterror.CombineErrors(errs)
+	}
+	return droppedRecords, nil
+}
+
+// appendAndSend appends line to the request body that will be sent and sends
+// the accumulated data if the internal buffer has been filled (with maxBufferSize elements).
+// It returns appendResponse
+func (s *sender) appendAndSend(
+	line string,
+	pipeline PipelineType,
+	body *strings.Builder,
+	flds fields,
+) appendResponse {
+	var errors []error
+	ar := newAppendResponse()
+
+	if body.Len() > 0 && body.Len()+len(line) >= s.config.MaxRequestBodySize {
+		ar.sent = true
+		if err := s.send(pipeline, strings.NewReader(body.String()), flds); err != nil {
+			errors = append(errors, err)
+		}
+		body.Reset()
```

Comment:
I know I mentioned it before, but just to be extra cautious: if you have concurrent calls to `appendAndSend` passing the same `body`, you'll have a concurrency issue here, as you are changing the parameter you are receiving without any sort of locks. If you are sure that this will only be called serially for each `body`, then it's fine, but it does look like you might have this buffer shared for your component as a whole.